lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp

```diff
@@ -604,103 +648,83 @@ void mlir::torch::onnx_c::populateDefaultDomainQtoZ(
-        // situation.
-        auto sortIndicesType = axesType.getWithSizesAndDtype(
-            axesType.getOptionalSizes(),
-            IntegerType::get(binder.op->getContext(), 64, IntegerType::Signed));
-        auto sortOpResult = rewriter.create<Torch::AtenSortOp>(
-            binder.getLoc(), axes.getType(), sortIndicesType, updatedAxes, zero,
-            cstFalse);
-        Value result;
-        auto baseType = Torch::ValueTensorType::getWithLeastStaticInformation(
-            binder.op->getContext());
-        // Go through the updated, sorted axes. Do unsqueeze for each dim.
-        for (int i = 0; i < sizes[0]; i++) {
-          Value selectIndex = rewriter.create<Torch::ConstantIntOp>(
-              binder.getLoc(), rewriter.getType<Torch::IntType>(),
-              rewriter.getIntegerAttr(rewriter.getIntegerType(64), i));
-          Value extract = rewriter.create<Torch::AtenSelectIntOp>(
-              binder.getLoc(), selectResultType, sortOpResult->getResult(0),
-              zero, selectIndex);
-          Value dim = rewriter.create<Torch::AtenItemOp>(
-              binder.getLoc(), rewriter.getType<Torch::IntType>(), extract);
-          if (sizes[0] == 1) {
-            result = rewriter.create<Torch::AtenUnsqueezeOp>(
-                binder.getLoc(), resultType, data, dim);
-          } else if (i == 0) {
-            result = rewriter.create<Torch::AtenUnsqueezeOp>(
-                binder.getLoc(), baseType, data, dim);
-          } else if (i == sizes[0] - 1) {
-            result = rewriter.create<Torch::AtenUnsqueezeOp>(
-                binder.getLoc(), resultType, result, dim);
-          } else {
-            result = rewriter.create<Torch::AtenUnsqueezeOp>(
-                binder.getLoc(), baseType, result, dim);
+
+        SmallVector<int64_t> unsqueezeDims;
+        SmallVector<int64_t> inputShape(inputType.getSizes());
+        if (inputType.areAllSizesKnown() && resultType.areAllSizesKnown()) {
+          // If the input shape and result shape is statically known then the
+          // list of dims to be squeezed can be derived from those shapes. As a
+          // result, we don't have to wait for the dim values to be known at
+          // runtime which is also expected by the downstream pipeline.
+          SmallVector<int64_t> resultShape(resultType.getSizes());
+          unsigned inputShapeCounter = 0;
+          for (unsigned i = 0; i < resultRank; i++) {
+            if (inputShapeCounter < inputRank &&
+                inputShape[inputShapeCounter] == resultShape[i]) {
+              inputShapeCounter++;
+            } else {
+              unsqueezeDims.push_back(i);
+            }
+          }
+        } else {
+          auto constOp = axes.getDefiningOp<Torch::OperatorOp>();
+          if (!constOp || !constOp.getName().equals("onnx.Constant"))
+            return failure();
+          DenseResourceElementsAttr attr =
```

Comment:
You should not replicate grabbing the constant value by assuming the input is an onnx constant operation. You should be able to use the `m_Constant` to fold. This runs via the dialect conversion framework so I am reasonably certain it should be able to pull the constant values that way.

Reply:
I tried using that but it doesn't work for the `DenseResourceElementsAttr`.

Reply:
Hi @rsuderman, let me know if we can proceed with this patch for now since it has been pending since long.

Reply:
If thats the case then we should make a constant matcher for the `onnx.Constant` operation instead of embedding the logic within the rewriter. See examples of how constant folding matchers are written for torch constants:

https://github.com/llvm/torch-mlir/blob/ce7d4f1660c210b98978b73a0ed4a3db440e4646/include/torch-mlir/Dialect/Torch/IR/TorchOps.h#L140

We should also guarantee the resource variable matching occurs using the same code for `onnx.Constant`. There is a reasonable chance we will rework constant loading at some point and need to handle alternative methods for loading constants.

Reply:
As far as I collected from the onnx documentation, you shouldn't need to match on constant axes, because even with dynamic dims on the input, the onnx shape inference for this op checks if the input axes can be realized (e.g. if they are coming from a constant), and adds the 1's to where they need to be. See, for example, the generated onnx graph for ElementwisePreluModule_basic and [onnx.Unsqueeze](https://github.com/onnx/onnx/blob/366ac64b5af0af2d02b019d84664e6b46b3ceb09/onnx/defs/tensor/defs.cc#L1849). That is, the check on line 676 is not entirely necessary for the logic of that block to work,